### PR TITLE
Implement HTTP header handling properly

### LINKED ISSUE #105: Revisit HTTP header encoding
Currently struggling to understand the spec. See https://github.com/cloudevents/spec/issues/777

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #105

This implements the spec clarification in
https://github.com/cloudevents/spec/pull/793, so we mustn't merge
this until that's been approved and merged.

Signed-off-by: Jon Skeet <jonskeet@google.com>

### CODE REVIEW

File: src/CloudNative.CloudEvents/Http/HttpUtilities.cs [hunk 1]
```diff
@@ -69,42 +70,242 @@ public static class HttpUtilities
...
+                    builder.Append((char) codePoint);
+                    continue;
+                }
+
+                // Full-on UTF-8 encoding now...
+                if (buffer is null)
+                {
+                    buffer = new byte[4];
+                }
+                int byteCount = Encoding.UTF8.GetBytes(value, i, codePoint > 0xffff ? 2 : 1, buffer, 0);
```
Comment (JoshLove-msft) [inline src/CloudNative.CloudEvents/Http/HttpUtilities.cs:121]: Do we have a test for the case where the codePoint > 0xffff?

File: src/CloudNative.CloudEvents/Http/HttpUtilities.cs [hunk 2]
```diff
@@ -69,42 +70,242 @@ public static class HttpUtilities
...
+                for (int j = 0; j < byteCount; j++)
+                {
+                    builder.Append('%');
+                    builder.Append(HexDigits[buffer[j] >> 4]);
+                    builder.Append(HexDigits[buffer[j] & 0x0f]);
+                }
+
+                // Ensure we iterate over code points, not UTF-16 code units.
+                if (codePoint > 0xffff)
```
Comment (JoshLove-msft) [inline src/CloudNative.CloudEvents/Http/HttpUtilities.cs:130]: Would it be worth consolidating this check with the one on line 121.

File: src/CloudNative.CloudEvents/Http/HttpUtilities.cs [hunk 3]
```diff
@@ -69,42 +70,242 @@ public static class HttpUtilities
...
+            for (int i = 0; i < value.Length; i++)
+            {
+                char c = value[i];
+                if (c != '%')
+                {
+                    builder.Append(c);
+                    continue;
+                }
+
+                // Percent encoding is remarkably complex to handle correctly and efficiently.
```
Comment (JoshLove-msft) [inline src/CloudNative.CloudEvents/Http/HttpUtilities.cs:165]: So I see 😄

File: test/CloudNative.CloudEvents.UnitTests/Http/HttpUtilitiesTest.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    public class HttpUtilitiesTest
+    {
+        [Theory]
+        [InlineData("simple", "simple")]
+        [InlineData("Euro \u20AC \U0001F600", "Euro%20%E2%82%AC%20%F0%9F%98%80")]
+        [InlineData("space encoded", "space%20encoded")]
+        [InlineData("percent%encoded", "percent%25encoded")]
+        [InlineData("quote\"encoded", "quote%22encoded")]
+        [InlineData("caf\u00e9", "caf%C3%A9")]
+        public void RoundTripHeaderValue(string original, string encoded)
```
Comment (JoshLove-msft) [inline test/CloudNative.CloudEvents.UnitTests/Http/HttpUtilitiesTest.cs:19]: Would it be worth having one-way Encode/Decode tests as well?

**Review by JoshLove-msft — APPROVED** [review]

File: src/CloudNative.CloudEvents/Http/HttpUtilities.cs (hunk 1 above)
Comment (jskeet) [reply src/CloudNative.CloudEvents/Http/HttpUtilities.cs:121]: Yes, we have a test that includes an emoji.

File: src/CloudNative.CloudEvents/Http/HttpUtilities.cs (hunk 2 above)
Comment (jskeet) [reply src/CloudNative.CloudEvents/Http/HttpUtilities.cs:130]: Done.

File: test/CloudNative.CloudEvents.UnitTests/Http/HttpUtilitiesTest.cs (hunk 4 above)
Comment (jskeet) [reply test/CloudNative.CloudEvents.UnitTests/Http/HttpUtilitiesTest.cs:19]: I started with those, but then found there was just a lot of duplication.

While we could extract the test cases into TheoryData and have two tests that both use the same TheoryData member, it felt simpler to just combine them.

**Review by jskeet — COMMENTED** [review]
Will include the small change from review comments in a rebased version. Thanks for the review! If the spec is approved in the WG meeting later, I'll merge.

**Pull request merged**